# HR Analytics Project- Understanding the Attrition in HR

### Problem Statement:
Every year a lot of companies hire a number of employees. The companies invest time and money in training those employees, not just this but there are training programs within the companies for their existing employees as well. The aim of these programs is to increase the effectiveness of their employees. But where HR Analytics fit in this? and is it just about improving the performance of employees?

### HR Analytics

Human resource analytics (HR analytics) is an area in the field of analytics that refers to applying analytic processes to the human resource department of an organization in the hope of improving employee performance and therefore getting a better return on investment. HR analytics does not just deal with gathering data on employee efficiency. Instead, it aims to provide insight into each process by gathering data and then using it to make relevant decisions about how to improve these processes.

### Attrition in HR

Attrition in human resources refers to the gradual loss of employees overtime. In general, relatively high attrition is problematic for companies. HR professionals often assume a leadership role in designing company compensation programs, work culture, and motivation systems that help the organization retain top employees.

How does Attrition affect companies? and how does HR Analytics help in analyzing attrition? We will discuss the first question here and for the second question, we will write the code and try to understand the process step by step.

### Attrition affecting Companies

A major problem in high employee attrition is its cost to an organization. Job postings, hiring processes, paperwork, and new hire training are some of the common expenses of losing employees and replacing them. Additionally, regular employee turnover prohibits your organization from increasing its collective knowledge base and experience over time. This is especially concerning if your business is customer-facing, as customers often prefer to interact with familiar people. Errors and issues are more likely if you constantly have new workers.

# Importing Libraries

In [3]:
# To Read and Process Data
import pandas as pd
import numpy as np


# For data Visualization
import seaborn as sns
import matplotlib.pyplot as plt

# Getting over warning messages
import warnings
warnings.filterwarnings('ignore')

# For Encoding Categorical Data
from sklearn.preprocessing import LabelEncoder

# for scaling
from sklearn.preprocessing import StandardScaler

# To display all columns
pd.pandas.set_option('display.max_columns',None)

# -------------------------------Getting to Know About Data---------------------------------------


# Reading File

In [4]:
df = pd.read_csv('WA_Fn-UseC_-HR-Employee-Attrition.csv')

## 1. Overall Data Analysis

In [8]:
# getting to know size of data set, to know overall records, and columns
print(f'Number of rows and columns in given Data Frame is {df.shape}')

Number of rows and columns in given Data Frame is (1470, 35)


In [6]:
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  \
0                        2  Female          94               3         2   
1                        3    Male          61               2         2   
2                        4    Male          92               2         1   
3                        4  Female          56               3         1   
4                        1    Male          40               3         1   

                 JobRole  JobSatisfaction MaritalStatus  MonthlyIncome  \
0        Sales Executive                4        Single           5993   
1     Research Scientist                2       Married           5130   
2  Laboratory Technician                3        Single           2090   
3     Research Scientist                3       Married           2909   
4  Laboratory Technician                2       Married           3468   

   MonthlyRate  NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
0        19479                   8      Y      Yes                 11   
1        24907                   1      Y       No                 23   
2         2396                   6      Y      Yes                 15   
3        23159                   1      Y      Yes                 11   
4        16632                   9      Y       No                 12   

   PerformanceRating  RelationshipSatisfaction  StandardHours  \
0                  3                         1             80   
1                  4                         4             80   
2                  3                         2             80   
3                  3                         3             80   
4                  3                         4             80   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                 0                  8                      0   
1                 1                 10                      3   
2                 0                  7                      3   
3                 0                  8                      3   
4                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

In [9]:
df.tail()

Age Attrition     BusinessTravel  DailyRate              Department  \
1465   36        No  Travel_Frequently        884  Research & Development   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1468   49        No  Travel_Frequently       1023                   Sales   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
1465                23          2        Medical              1   
1466                 6          1        Medical              1   
1467                 4          3  Life Sciences              1   
1468                 2          3        Medical              1   
1469                 8          3        Medical              1   

      EmployeeNumber  EnvironmentSatisfaction Gender  HourlyRate  \
1465            2061                        3   Male          41   
1466            2062                        4   Male          42   
1467            2064                        2   Male          87   
1468            2065                        4   Male          63   
1469            2068                        2   Male          82   

      JobInvolvement  JobLevel                    JobRole  JobSatisfaction  \
1465               4         2      Laboratory Technician                4   
1466               2         3  Healthcare Representative                1   
1467               4         2     Manufacturing Director                2   
1468               2         2            Sales Executive                2   
1469               4         2      Laboratory Technician                3   

     MaritalStatus  MonthlyIncome  MonthlyRate  NumCompaniesWorked Over18  \
1465       Married           2571        12290                   4      Y   
1466       Married           9991        21457                   4      Y   
1467       Married           6142         5174                   1      Y   
1468       Married           5390        13243                   2      Y   
1469       Married           4404        10228                   2      Y   

     OverTime  PercentSalaryHike  PerformanceRating  RelationshipSatisfaction  \
1465       No                 17                  3                         3   
1466       No                 15                  3                         1   
1467      Yes                 20                  4                         2   
1468       No                 14                  3                         4   
1469       No                 12                  3                         1   

      StandardHours  StockOptionLevel  TotalWorkingYears  \
1465             80                 1                 17   
1466             80                 1                  9   
1467             80                 1                  6   
1468             80                 0                 17   
1469             80                 0                  6   

      TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
1465                      3                3               5   
1466                      5                3               7   
1467                      0                3               6   
1468                      3                2               9   
1469                      3                4               4   

      YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
1465                   2                        0                     3  
1466                   7                        1                     7  
1467                   2                        0                     3  
1468                   6                        0                     8  
1469                   3                        1                     2

In [10]:
df.sample(10)

Age Attrition     BusinessTravel  DailyRate              Department  \
1070   28        No  Travel_Frequently        467                   Sales   
291    36        No      Travel_Rarely        506  Research & Development   
1179   34        No      Travel_Rarely       1130  Research & Development   
1364   28        No  Travel_Frequently        783                   Sales   
765    38        No  Travel_Frequently       1186  Research & Development   
547    42       Yes  Travel_Frequently        933  Research & Development   
1235   46        No      Travel_Rarely       1277                   Sales   
677    49        No      Travel_Rarely        527  Research & Development   
312    31        No      Travel_Rarely        192  Research & Development   
954    42        No         Non-Travel        495  Research & Development   

      DistanceFromHome  Education    EducationField  EmployeeCount  \
1070                 7          3     Life Sciences              1   
291                  3          3  Technical Degree              1   
1179                 3          3     Life Sciences              1   
1364                 1          2     Life Sciences              1   
765                  3          4             Other              1   
547                 19          3           Medical              1   
1235                 2          3     Life Sciences              1   
677                  8          2             Other              1   
312                  2          4     Life Sciences              1   
954                  2          1     Life Sciences              1   

      EmployeeNumber  EnvironmentSatisfaction  Gender  HourlyRate  \
1070            1507                        3    Male          55   
291              397                        3    Male          30   
1179            1658                        4  Female          66   
1364            1927                        3    Male          42   
765             1060                        3    Male          44   
547              752                        3    Male          57   
1235            1732                        3    Male          74   
677              944                        1  Female          51   
312              426                        3    Male          32   
954             1334                        3    Male          37   

      JobInvolvement  JobLevel                JobRole  JobSatisfaction  \
1070               3         2        Sales Executive                1   
291                3         2     Research Scientist                2   
1179               3         2     Research Scientist                2   
1364               2         2        Sales Executive                4   
765                3         1     Research Scientist                3   
547                4         1     Research Scientist                3   
1235               3         3        Sales Executive                4   
677                3         3  Laboratory Technician                2   
312                3         1     Research Scientist                4   
954                3         4                Manager                3   

     MaritalStatus  MonthlyIncome  MonthlyRate  NumCompaniesWorked Over18  \
1070        Single           4898        11827                   0      Y   
291         Single           4485        26285                   4      Y   
1179      Divorced           5433        19332                   1      Y   
1364       Married           6834        19255                   1      Y   
765        Married           2821         2997                   3      Y   
547       Divorced           2759        20366                   6      Y   
1235      Divorced          10368         5596                   4      Y   
677        Married           7403        22477                   4      Y   
312       Divorced           2695         7747                   0      Y   
954        Married          17861        2

In [11]:
df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

### Observation - 
1. There are total 1470 records with 35 columns in each entry.
2. There are total 35 columns as below - 
  - Age: 
  - Attrition: 
  - BusinessTravel: 
  - DailyRate: 
  - Department: 
  - DistanceFromHome: 
  - Education: 
  - EducationField: 
  - EmployeeCount: 
  - EmployeeNumber: 
  - EnvironmentSatisfaction: 
  - Gender: 
  - HourlyRate: 
  - JobInvolvement: 
  - JobLevel: 
  - JobRole: 
  - JobSatisfaction: 
  - MaritalStatus: 
  - MonthlyIncome: 
  - MonthlyRate: 
  - NumCompaniesWorked: 
  - Over18: 
  - OverTime: 
  - PercentSalaryHike: 
  - PerformanceRating: 
  - RelationshipSatisfaction: 
  - StandardHours: 
  - StockOptionLevel: 
  - TotalWorkingYears: 
  - TrainingTimesLastYear: 
  - WorkLifeBalance: 
  - YearsAtCompany: 
  - YearsInCurrentRole: 
  - YearsSinceLastPromotion: 
  - YearsWithCurrManager: 

## 2. Getting to Know More About Data

In [12]:
df.dtypes

Age                          int64
Attrition                   object
BusinessTravel              object
DailyRate                    int64
Department                  object
DistanceFromHome             int64
Education                    int64
EducationField              object
EmployeeCount                int64
EmployeeNumber               int64
EnvironmentSatisfaction      int64
Gender                      object
HourlyRate                   int64
JobInvolvement               int64
JobLevel                     int64
JobRole                     object
JobSatisfaction              int64
MaritalStatus               object
MonthlyIncome                int64
MonthlyRate                  int64
NumCompaniesWorked           int64
Over18                      object
OverTime                    object
PercentSalaryHike            int64
PerformanceRating            int64
RelationshipSatisfaction     int64
StandardHours                int64
StockOptionLevel             int64
TotalWorkingYears   

# 3 Gettting to Know Five Number Summery for Continuous variable¶

In [24]:
df.describe().T

count          mean          std     min      25%  \
Age                       1470.0     36.923810     9.135373    18.0    30.00   
DailyRate                 1470.0    802.485714   403.509100   102.0   465.00   
DistanceFromHome          1470.0      9.192517     8.106864     1.0     2.00   
Education                 1470.0      2.912925     1.024165     1.0     2.00   
EmployeeCount             1470.0      1.000000     0.000000     1.0     1.00   
EmployeeNumber            1470.0   1024.865306   602.024335     1.0   491.25   
EnvironmentSatisfaction   1470.0      2.721769     1.093082     1.0     2.00   
HourlyRate                1470.0     65.891156    20.329428    30.0    48.00   
JobInvolvement            1470.0      2.729932     0.711561     1.0     2.00   
JobLevel                  1470.0      2.063946     1.106940     1.0     1.00   
JobSatisfaction           1470.0      2.728571     1.102846     1.0     2.00   
MonthlyIncome             1470.0   6502.931293  4707.956783  1009.0  2911.00   
MonthlyRate               1470.0  14313.103401  7117.786044  2094.0  8047.00   
NumCompaniesWorked        1470.0      2.693197     2.498009     0.0     1.00   
PercentSalaryHike         1470.0     15.209524     3.659938    11.0    12.00   
PerformanceRating         1470.0      3.153741     0.360824     3.0     3.00   
RelationshipSatisfaction  1470.0      2.712245     1.081209     1.0     2.00   
StandardHours             1470.0     80.000000     0.000000    80.0    80.00   
StockOptionLevel          1470.0      0.793878     0.852077     0.0     0.00   
TotalWorkingYears         1470.0     11.279592     7.780782     0.0     6.00   
TrainingTimesLastYear     1470.0      2.799320     1.289271     0.0     2.00   
WorkLifeBalance           1470.0      2.761224     0.706476     1.0     2.00   
YearsAtCompany            1470.0      7.008163     6.126525     0.0     3.00   
YearsInCurrentRole        1470.0      4.229252     3.623137     0.0     2.00   
YearsSinceLastPromotion   1470.0      2.187755     3.222430     0.0     0.00   
YearsWithCurrManager      1470.0      4.123129     3.568136     0.0     2.00   

                              50%       75%      max  
Age                          36.0     43.00     60.0  
DailyRate                   802.0   1157.00   1499.0  
DistanceFromHome              7.0     14.00     29.0  
Education                     3.0      4.00      5.0  
EmployeeCount                 1.0      1.00      1.0  
EmployeeNumber             1020.5   1555.75   2068.0  
EnvironmentSatisfaction       3.0      4.00      4.0  
HourlyRate                   66.0     83.75    100.0  
JobInvolvement                3.0      3.00      4.0  
JobLevel                      2.0      3.00      5.0  
JobSatisfaction               3.0      4.00      4.0  
MonthlyIncome              4919.0   8379.00  19999.0  
MonthlyRate               14235.5  20461.50  26999.0  
NumCompaniesWorked            2.0      4.00      9.0  
PercentSalaryHike            14.0     18.00     25.0  
PerformanceRating             3.0      3.00      4.0  
RelationshipSatisfaction      3.0      4.00      4.0  
StandardHours                80.0     80.00     80.0  
StockOptionLevel              1.0      1.00      3.0  
TotalWorkingYears            10.0     15.00     40.0  
TrainingTimesLastYear         3.0      3.00      6.0  
WorkLifeBalance               3.0      3.00      4.0  
YearsAtCompany                5.0      9.00     40.0  
YearsInCurrentRole            3.0      7.00     18.0  
YearsSinceLastPromotion       1.0      3.00     15.0  
YearsWithCurrManager          3.0      7.00     17.0

### Observations - 
Looking at count column, there are no missing values. There may be outliers, but no missing values in continuous data
1. Age: Average age of employees is 36
2. DailyRate: Average daily rate is 802.49 with min. rate 102 abd max. rate is 1499
3. DistanceFromHome:
    - For 50% of employees have office within 7 km.
    - Average distance emlpoyee need to travel to reach company is 9.19
4. Education: Education of employees
5. EmployeeCount: Employee count as 1
6. EmployeeNumber: Employee Number
7. EnvironmentSatisfaction:
    - Average employee satisfication is 2.72
    - 50% of employee has satisfication level less than 3
    - 25% of employee has satisfication level less than 2
8. HourlyRate: Average hourly rate of employee is 65.89
9. JobInvolvement: Average 2.729 with min 1 and max 4
10. JobLevel: 50% of employees has job level 1 and 2
11. JobSatisfaction: Average job satisfication is 2.72
12. MonthlyIncome:
    - Average monthly income is 6502.93
    - 50 % of employees has income less than 4919.0
13. MonthlyRate: Average monthly rate of employee is 14313.103401
14. NumCompaniesWorked: Record having employees with 0 to 9 companies
15. PercentSalaryHike:
    - Average salary hike is 15.20%
    - 50% of emplpyees received, 14 % Salary hike
    - Minimum salary hike is 11% and maximum 25%
16. PerformanceRating:
    - All have performance rating of 3 and 4
17. RelationshipSatisfaction: Average relation satisfaction is 2.71
18. StandardHours: 80 Hours, Constant for all employees
19. StockOptionLevel: Average = 0.79, min = 1 and max = 3
20. TotalWorkingYears:
    - Average working years are 11.27
    - min is 0 and max is 40 years
    - 25% employees has experiance less than 6
    - 25% employees has experiance between 6 and 10
21. TrainingTimesLastYear: Total training time last year
22. WorkLifeBalance: Graded between 1 to 4 with average of 2.76
23. YearsAtCompany: 
    - 25% of employee has 9 to 40 years of experiance in this company
24. YearsInCurrentRole: with average of 4.229
25. YearsSinceLastPromotion: Employee get promotion in every 2 years (Average)
26. YearsWithCurrManager: 25%-25% employees have spent 0-2 and 7-17 years with current manager

# 4 Gettting to Know about Categorical Variable

In [6]:
df.describe(include="O")

Attrition BusinessTravel              Department EducationField Gender  \
count       1470           1470                    1470           1470   1470   
unique         2              3                       3              6      2   
top           No  Travel_Rarely  Research & Development  Life Sciences   Male   
freq        1233           1043                     961            606    882   

                JobRole MaritalStatus Over18 OverTime  
count              1470          1470   1470     1470  
unique                9             3      1        2  
top     Sales Executive       Married      Y       No  
freq                326           673   1470     1054

### Observations - 
There are less than 10 categories in each of qualitative data. This information will help us to plot various plots to undersand data.
1. Attrition: There are only two categories (Yes/No): With mode of "No" having frequency of 1233, for Yes it counts as 337
2. BusinessTravel: Three category, with mode as "Travel_Rarely" with freq. as 1043
3. Department: Three departments, with mode of "Research & Development" with frequency of 961
4. EducationField: 6 Category, with max "Life Sciences" counted 606
5. Gender: 2 cate, with mode of male with 882 count.
6. JobRole: 9 job role, with more sale execuative (326)
7. MaritalStatus: Most of the men working are Marrieed (673)
8. Over18: As per rule/qualification required, all employees must be above 18 years.
9. OverTime: Most of employees are not doing overtime. Only 416 